component/newstorage/edv/restprovider.go

```diff
@@ -0,0 +1,620 @@
+	errInvalidQueryExpressionFormat = errors.New("invalid expression format. " +
+		"it must be in the following format: TagName:TagValue")
+)
+
+// Option allows for configuration of a RESTProvider.
+type Option func(opts *RESTProvider)
+
+// WithTLSConfig is an option that allows for the definition of a secured HTTP transport using a tls.Config instance.
+func WithTLSConfig(tlsConfig *tls.Config) Option {
+	return func(opts *RESTProvider) {
+		opts.restClient.httpClient.Transport = &http.Transport{TLSClientConfig: tlsConfig}
+	}
+}
+
+// WithHeaders option is for setting additional http request headers (since it's a function, it can call a remote
+// authorization server to fetch the necessary info needed in these headers).
+func WithHeaders(addHeadersFunc addHeaders) Option {
+	return func(opts *RESTProvider) {
+		opts.restClient.headersFunc = addHeadersFunc
+	}
+}
+
+// WithFullDocumentsReturnedFromQueries option is a performance optimization that speeds up queries by getting
+// full documents from the EDV server instead of only document locations - each of which would require a separate REST
+// call to retrieve. The EDV server that this RESTProvider connects to must support the TrustBloc EDV server extension
+// as defined here: https://github.com/trustbloc/edv/blob/main/docs/extensions.md#return-full-documents-on-query.
+func WithFullDocumentsReturnedFromQueries() Option {
+	return func(restProvider *RESTProvider) {
+		restProvider.returnFullDocumentsOnQuery = true
+	}
+}
+
+// WithBatchEndpointExtension option is a performance optimization that allows for restStore.Batch to only require one
+// REST call. The EDV server that this RESTProvider connects to must support the TrustBloc EDV server extension
+// as defined here: https://github.com/trustbloc/edv/blob/main/docs/extensions.md#batch-endpoint.
+func WithBatchEndpointExtension() Option {
+	return func(opts *RESTProvider) {
+		opts.batchEndpointExtensionEnabled = true
+	}
+}
+
+// RESTProvider is a newstorage.Provider that can be used to store data in a server supporting the
+// data vault HTTP API as defined in https://identity.foundation/confidential-storage/#http-api.
+type RESTProvider struct {
+	vaultID    string
+	formatter  *EncryptedFormatter
+	restClient *restClient
+	openStores map[string]*restStore
+	lock       sync.RWMutex
+
+	returnFullDocumentsOnQuery    bool
+	batchEndpointExtensionEnabled bool
+}
+
+// NewRESTProvider returns a new RESTProvider. edvServerURL is the base URL for the EDV server.
+// vaultID is the ID of the vault where this provider will store data. The vault must be created in advance, and since
+// the EDV REST API does not provide a method to check if a vault with a given ID exists, any errors due to a
+// non-existent vault will be deferred until calls are actually made to it in the store.
+func NewRESTProvider(edvServerURL, vaultID string, formatter *EncryptedFormatter, options ...Option) *RESTProvider {
```

Comment:
One major difference between this new RESTProvider implementation and the old one is that this new one now directly takes in an EncryptedFormatter instead of expecting to be used as an underlying provider in a format provider. This has made the code much cleaner - before the RESTProvider was somewhat confusing since the formatted provider and RESTProvider both needed MAC crypto objects passed in, which I always thought was kind of weird. Now if you want to use an EDV server, you just create a REST provider with an Encrypted Formatter.

The Encrypted Formatter can still be used with a formatted store if someone wanted to use the EDV data model with some other type of storage provider. (allowing for this use case was one of my original requirements, so I needed to maintain that functionality)